### PR TITLE
tests: fix "job canceled" issue and improve cleanup for snaps

### PR DESCRIPTION
**Pull request by sergiocazzolato**
Some issues are fixed/improved with change.

The first is how the snapd service is stopped. To avoid the error "Job
canceled" when the snapd service is stopped but it is in "activating"
state. To deal with that, the solution is to wait until the state change
to active, to stop it.

The second is how the services are stopped in the postrm scritps. The
idea of this change is to avoid call "systemctl stop service" many times
when the services take some time to be stopped. This was causing some
services like the lxd or neutron are stopped many times.

The third is that the postrm script will fail in case the final cleanup
is not completed properly.

The last once is a change in the install-snaps test to remove the snap
as part of the test and check the snap is properly removed.

### CODE REVIEW

**Comment by zyga** [comment]
This sequence makes no sense:
```
+ rmdir --ignore-fail-on-non-empty /var/snap
+ [ -d /var/snap ]
+ echo Failed cleaning dir /var/snap
Failed cleaning dir /var/snap
+ exit 1
```

We should at least print the files remaining in `/var/snap`.

File: tests/lib/reset.sh [hunk 1]
```diff
@@ -10,6 +10,20 @@ reset_classic() {
     # have changed on the disk.
     systemctl daemon-reload
 
+    if [[ "$SPREAD_SYSTEM" == ubuntu-14.04-* ]]; then
+        echo "Ensure the service is active before stop it"
```
Comment (stolowski) [inline tests/lib/reset.sh]: "stop" -> "stopping".

File: packaging/ubuntu-14.04/snapd.postrm [hunk 2]
```diff
@@ -4,13 +4,19 @@ set -e
 
 systemctl_stop() {
     unit="$1"
-    for i in $(seq 10); do
+    if ! systemctl is-active -q "$unit"; then
+        echo "$unit is stopped."
+        return
```
Comment (zyga) [inline packaging/ubuntu-14.04/snapd.postrm]: The unit may become active at this time and we will just not catch that.

Could we always attempt to stop it and then continue to the waiting loop below?

File: packaging/ubuntu-14.04/snapd.postrm [hunk 3]
```diff
@@ -74,8 +80,14 @@ if [ "$1" = "purge" ]; then
 
     echo "Final directory cleanup"
     for d in "/snap/bin" "/snap" "/var/snap"; do
+        # Force remove due to directories for old revisions could still exist
+        if [ -d "$d" ]; then
+            rm -rf  $d
```
Comment (zyga) [inline packaging/ubuntu-14.04/snapd.postrm]: Please quote `"$d"`

File: packaging/ubuntu-14.04/snapd.postrm [hunk 4]
```diff
@@ -74,8 +80,14 @@ if [ "$1" = "purge" ]; then
...
+        fi
         if [ -d "$d" ]; then
-            rmdir --ignore-fail-on-non-empty $d
+            echo "Failed cleaning dir $d"
```
Comment (zyga) [inline packaging/ubuntu-14.04/snapd.postrm]: Nitpick: `cannot remove directory $d`

File: packaging/ubuntu-16.04/snapd.postrm [hunk 5]
```diff
@@ -77,8 +83,14 @@ if [ "$1" = "purge" ]; then
 
     echo "Final directory cleanup"
     for d in "/snap/bin" "/snap" "/var/snap"; do
+        # Force remove due to directories for old revisions could still exist
+        if [ -d "$d" ]; then
+            rm -rf  $d
```
Comment (zyga) [inline packaging/ubuntu-16.04/snapd.postrm]: Ditto, (I'd love to de-duplicate this)

File: packaging/ubuntu-16.04/snapd.postrm [hunk 6]
```diff
@@ -77,8 +83,14 @@ if [ "$1" = "purge" ]; then
...
+        fi
         if [ -d "$d" ]; then
-            rmdir --ignore-fail-on-non-empty $d
+            echo "Failed cleaning dir $d"
```
Comment (zyga) [inline packaging/ubuntu-16.04/snapd.postrm]: Ditto

File: tests/lib/reset.sh [hunk 7]
```diff
@@ -10,6 +10,20 @@ reset_classic() {
     # have changed on the disk.
     systemctl daemon-reload
 
+    if [[ "$SPREAD_SYSTEM" == ubuntu-14.04-* ]]; then
```
Comment (zyga) [inline tests/lib/reset.sh]: Why is this special cased for Ubuntu 14.04?

Comment (sergiocazzolato) [reply tests/lib/reset.sh]: @zyga, I could not reproduce that error in other releases, also all the failures that we see are always happening just on 14.04., I guess it is a problem on systemd and just happening on that version. I can make it independent of the OS, that should not affect.

Comment (zyga) [reply tests/lib/reset.sh]: I'd prefer if it was universally enabled as all the special casing makes it harder to keep a mental model of what is going on where. If it works let's enable it everywhere.

Comment (sergiocazzolato) [reply tests/lib/reset.sh]: @zyga addressed in the last commit, thanks for reviewing this.

File: packaging/ubuntu-14.04/snapd.postrm [hunk 8]
```diff
@@ -4,13 +4,16 @@ set -e
 
 systemctl_stop() {
     unit="$1"
-    for i in $(seq 10); do
+
+    echo "Stoping $unit [attempt $i]"
```
Comment (zyga) [inline packaging/ubuntu-14.04/snapd.postrm]: `$i` is no longer a variable here, perhaps you meant just `echo "Stopping $unit"`

File: packaging/ubuntu-16.04/snapd.postrm [hunk 9]
```diff
@@ -4,13 +4,16 @@ set -e
 
 systemctl_stop() {
     unit="$1"
-    for i in $(seq 10); do
+
+    echo "Stoping $unit [attempt $i]"
```
Comment (zyga) [inline packaging/ubuntu-16.04/snapd.postrm]: Same as above wrt `$i`

File: packaging/ubuntu-14.04/snapd.postrm [hunk 10]
```diff
@@ -74,8 +77,14 @@ if [ "$1" = "purge" ]; then
 
     echo "Final directory cleanup"
     for d in "/snap/bin" "/snap" "/var/snap"; do
+        # Force remove due to directories for old revisions could still exist
+        if [ -d "$d" ]; then
+            rm -rf  "$d"
```
Comment (zyga) [inline packaging/ubuntu-14.04/snapd.postrm]: Nitpick: two spaces between `-rf` and `"$d"`

File: packaging/ubuntu-16.04/snapd.postrm [hunk 11]
```diff
@@ -77,8 +80,14 @@ if [ "$1" = "purge" ]; then
 
     echo "Final directory cleanup"
     for d in "/snap/bin" "/snap" "/var/snap"; do
+        # Force remove due to directories for old revisions could still exist
+        if [ -d "$d" ]; then
+            rm -rf  "$d"
```
Comment (zyga) [inline packaging/ubuntu-16.04/snapd.postrm]: Same as above

**Comment by sergiocazzolato** [comment]
@zyga , thanks for reviewing, comments addressed, sorry for those errors, I should pay more attention doing these simple changes

File: packaging/ubuntu-14.04/snapd.postrm [hunk 12]
```diff
@@ -74,8 +77,14 @@ if [ "$1" = "purge" ]; then
...
     for d in "/snap/bin" "/snap" "/var/snap"; do
+        # Force remove due to directories for old revisions could still exist
+        if [ -d "$d" ]; then
+            rm -rf "$d"
+        fi
         if [ -d "$d" ]; then
-            rmdir --ignore-fail-on-non-empty $d
+            echo "Cannot remove directory $d"
+            ls "$d"
+            exit 1
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply packaging/ubuntu-14.04/snapd.postrm]: Ignore this comment please. I had reviewed the PR and added some commits last week, but actually never sent out the review. I see that this if fixed already in your latest commit.

**Pull request merged**